raiden/storage/restore.py

```diff
@@ -0,0 +1,80 @@
+from contextlib import contextmanager
+from copy import deepcopy
+
+from raiden.transfer import node, views
+from raiden.transfer.architecture import StateManager
+from raiden.transfer.state import NettingChannelState
+from raiden.transfer.utils import hash_balance_data
+from raiden.utils import typing
+
+
+@contextmanager
+def temporary_state_manager(state_transition, state):
+    state_manager = StateManager(state_transition, state)
+    try:
+        yield state_manager
+    finally:
+        state_manager = None
+
+
+def channel_state_until_balance_hash(
+        raiden: 'RaidenService',
+        token_address: typing.TokenAddress,
+        channel_identifier: typing.ChannelID,
+        target_balance_hash: bytes,
+) -> typing.Optional[NettingChannelState]:
+    """ Go through WAL state changes until a certain balance hash is found. """
+
+    # Restore state from the latest snapshot
+    snapshot = raiden.wal.storage.get_latest_state_snapshot()
+    if not snapshot:
+        # No snapshots were taken before this taking place.
+        # Therefore, we return a copy of the current channel state
+        channel_state = deepcopy(views.get_channelstate_by_id(
+            chain_state=views.state_from_raiden(raiden),
+            payment_network_id=raiden.default_registry.address,
+            token_address=token_address,
+            channel_id=channel_identifier,
+        ))
+        return channel_state
```

Comment:
this is not correct, it's ignoring all the states from state change 0 until now, you should just start with an empty state and replay the state changes, like this:

https://github.com/raiden-network/raiden/blob/51ffd61082dcdb5ada484c867fb572bcf8d95329/raiden/storage/wal.py#L20-L27